CI - Minimal and full CI matrix impl (#2051)
valkey-io/valkey-glide: .github/DEVELOPER.md

#### Engine Matrix (engine-matrix.json)

Defines the versions of Valkey engine to test against:

```json
[
    { "type": "valkey", "version": "7.2.5", "run": "always" },
    { "type": "valkey", "version": "redis-7.0.15" },
    { "type": "valkey", "version": "redis-6.2.14", "run": "always" }
]
```

#### Build Matrix (build-matrix.json)

Defines the host environments for testing:

```json
[
    {
        "OS": "ubuntu",
        "RUNNER": "ubuntu-latest",
        "TARGET": "x86_64-unknown-linux-gnu",
        "run": ["always", "python", "node", "java"]
    }
    // ... other configurations
]
```

#### Adding New Versions or Hosts

To add a new engine version, update engine-matrix.json.
To add a new host environment, update build-matrix.json.

Ensure new entries have all required fields:

For engine versions: `type`, `version`, `run` (optional)
For hosts: `OS`, `RUNNER`, `TARGET`, `run`

#### Triggering Workflows

Push to main or create a pull request to run workflows automatically.
Add `Core changes` label to a pull request to trigger full matrix tests.
Use workflow_dispatch for manual triggers.

### Mutual vs. Language-Specific Components

#### Mutual

`Engine matrix`

> `Build matrix`

Author: Host is build matrix

# CI/CD Workflow Guide

> TODO: Add a description of the CI/CD workflow and its components.

Author: I'll react to all comments when rewriting, but specifically this one - i want all the CI/CD to be documented, all the components.

Currently the doc covers just the "main" workflows of pr and push, i want also release, build, rc, rc tests, ort and so on to be documented.

Devs touching the code base a lot, and on each component there are more than enough devs familiar with it, so even if there's a miss here and there with readability or docs that's fine.

In the case of the CI/CD, there's 2-3 that has a really good familiarity with all the components.
The good scenario is 2 devs become DevOps against their will, the bad scenario is 1-2 leaves, and one is sick while there's a Hot fix that needs to be released immediately and nobody knows what to do.

So...

# CI/CD Workflow Guide

TODO: Add a description of the CI/CD workflow and its components.

> ### Overview

Author: The file need to be edited for fitting the last changes with scheduling.

Author: Maybe we should add a short component on the specific schedule WF

        "always-run-versions": ["17"]
    },
    // ... other configurations
]
```
* `language`: The language for which the version is supported.
* `versions`: The full versions supported of the language which will test against scheduled.
* `always-run-versions`: The versions that will always be tested, regardless of the workflow trigger.

#### Triggering Workflows

- Push to `main` by merging a PR or create a new pull request to run workflows automatically.
- Use `workflow_dispatch` for manual triggers, accepting a boolean configuration parameter to run all configurations.
- Scheduled runs are triggered daily to ensure regular testing of all configurations.

### Mutual vs. Language-Specific Components

#### Mutual

- Matrix files - `.github/json_matrices/`
- Shared dependencies installation - `.github/workflows/install-shared-dependencies/action.yml`
- Rust linters - `.github/workflows/lint-rust/action.yml`

#### Language-Specific

- Package manager commands
- Testing frameworks
- Build processes

### Customizing Workflows

Modify `<language>.yml` files to adjust language-specific steps.
Update matrix files to change tested versions or environments.

> Adjust cron schedules in workflow files for different timing of scheduled runs.

Author: for example here

# CI/CD Workflow Guide

TODO: Add a description of the CI/CD workflow and its components.

### Overview

Our CI/CD pipeline tests and builds our project across multiple languages, versions, and environments. This guide outlines the key components and processes of our workflow.

### Workflow Triggers

*   Push to `main` branch
*   Pull requests
*   Scheduled runs (daily)
*   Manual trigger (workflow_dispatch)
*   Pull request with `Core changes` label

### Language-Specific Workflows

* Python (python.yml)

* Node.js (node.yml)

* Java (java.yml)

Each language has its own workflow file with similar structure but language-specific steps.

### Shared Components

#### Run Field

> The `run` field in both engine-matrix.json and build-matrix.json is used to specify when a particular configuration should be executed:

Reviewer: I would start stating that we store the matrices that we load to the workflows under the json_matrices folder

> *   In engine*matrix.json:

Reviewer: ```suggestion
*   In engine-matrix.json:
```

    *   `"run": "always"`: The engine version will be tested in every workflow run.
    *   If `run` is omitted: The engine version will only be tested in full matrix runs (e.g., when the `Core changes` label is applied).

> *   In build*matrix.json:

Reviewer: ```suggestion
*   In build-matrix.json:
```

    *   The `run` array specifies which language workflows should use this host configuration.
    *   `"always"` in the array means the host will be used in every workflow run for the specified languages.

This allows for flexible control over which configurations are tested in different scenarios, optimizing CI/CD performance and resource usage.

#### Engine Matrix (engine-matrix.json)

Defines the versions of Valkey engine to test against:

```json
[
    { "type": "valkey", "version": "7.2.5", "run": "always" },
    { "type": "valkey", "version": "redis-7.0.15" },
    { "type": "valkey", "version": "redis-6.2.14", "run": "always" }
]
```

#### Build Matrix (build-matrix.json)

> Defines the host environments for testing:

Reviewer: same - please define what each parameter is used for

```json
[
    {
        "OS": "ubuntu",
        "RUNNER": "ubuntu-latest",
        "TARGET": "x86_64-unknown-linux-gnu",
        "run": ["always", "python", "node", "java"]
    }
    // ... other configurations
]
```

#### Adding New Versions or Hosts

To add a new engine version, update engine-matrix.json.
To add a new host environment, update build-matrix.json.

Ensure new entries have all required fields:

For engine versions: `type`, `version`, `run` (optional)

> For hosts: `OS`, `RUNNER`, `TARGET`, `run`

Reviewer: same - would cost extra maintainance to add the exact fields here

#### Triggering Workflows

Push to main or create a pull request to run workflows automatically.
Add `Core changes` label to a pull request to trigger full matrix tests.
Use workflow_dispatch for manual triggers.

### Mutual vs. Language-Specific Components

#### Mutual

`Engine matrix`
`Build matrix`
`Shared dependencies installation`

> `Linting (Rust)`

Reviewer: please use the file names - otherwise it isn't clear what you're referring to

> #### Language*Specific

Reviewer: ```suggestion
#### Language-Specific
```

`Package manager commands`
`Testing frameworks`
`Build processes`

> `Version matrices`

Reviewer: same

Reviewer: actually i'm not sure what is it 
do you refer to files here? to parameter? isn't clear..

Author: Components in workflows, would not add files here since it's language specific and will need to be maintained + too long and messy and not worth the gain
If you think it's redundant I may agree, not sure how needed it is.

### Customizing Workflows

Modify `[language].yml` files to adjust language-specific steps.
Update matrix files to change tested versions or environments.

> Adjust cron schedules in workflow files for different timing of scheduled runs.

Reviewer: make sure to remove it if you find a better way to do the daily schedule rather than specifying the exact hour on each

Author: Specific hours are like specifying specific languages, it is extra maintenance, why it is different here?

BTW didn't find better method, other running all the scheduled in one file, hence can control the execution order in a more deterministic way.

But while our method can result in some parallel running in some unusual occasion, I'm not sure why it's a problem.
Except for the self-hosted runner they can all run together in peace, and for the self-host it might just be slowing the progress but no real harm, and anyhow one hour suppose to be more than enough for specific language workflow.

### Workflow Matrices

> We use dynamic matrices for our CI/CD workflows, which are created using the `create-tests-matrices` action. This action is defined in `.github/workflows/create-tests-matrices/action.yml`.

Reviewer: nit: rename `create-tests-matrices` to `create-test-matrices`

> We use dynamic matrices for our CI/CD workflows, which are created using the `create-tests-matrices` action. This action is defined in `.github/workflows/create-tests-matrices/action.yml`.

Reviewer: can you elaborate more whats the propose of this action, why we use it

Author: Does text at the end of the section not sufficient? I feel like I'm repeating the usage idea twice already

#### How it works

1. The action is called with a `language-name` input.
2. It checks for the presence of a "Core changes" label on pull requests.
3. It generates two matrices:

>    - Engine matrix: Defines the versions of Valkey engine to test against.

Reviewer: ```suggestion
   - Engine matrix: Defines the types and versions of the engine to test against.
```

>    - Host matrix: Defines the host environments for testing.

Reviewer: ```suggestion
   - Host matrix: Defines the host platforms to run the tests on, for example Ubuntu on ARM64.
```

#### Outputs

- `should_run`: Indicates whether the job should run based on the presence of the "Core changes" label.
- `engine-matrix-output`: The generated engine matrix.
- `host-matrix-output`: The generated host matrix.

#### Matrix Generation Logic

- For pull requests and pushes, only configurations marked with `"run": "always"` are included in the matrices.
- For other events (like scheduled runs), all configurations are included.
- The host matrix is filtered based on the `language-name` input and the `run` array in the build-matrix.json file.

#### Customizing Matrices

- To modify the engine versions tested, update the `.github/json_matrices/engine-matrix.json` file.

> - To change the host environments, modify the `.github/json_matrices/build-matrix.json` file.

Reviewer: you already wrote in on lines 74-75:
> To add a new engine version, update engine-matrix.json.
To add a new host environment, update build-matrix.json.
